testcontainers/src/compose/mod.rs

```diff
@@ -0,0 +1,348 @@
+//! Testing with Docker compose
+//!
+//! Requirements:
+//!
+//! * enabling the `compose` feature.
+//! * have a `docker compose` v2.x.x installed
+//!
+//! ```
+//! let dc = DockerCompose::builder()
+//!     .env("MY_ENV_VAR", "MyValue")
+//!     .wait("mydb", WaitFor::message_on_stdout("database is ready"))
+//!     .wait("webserver", WaitFor::Healthcheck)
+//!     .build();
+//! dc.up();
+//! dc.block_until_ready();
+//!
+//! let port = dc.get_mapped_port("webserver", 80).unwrap();
+//! // testing my web server...
+//! ```
+//! You can configure the behavior when the `DockerCompose` is dropped
+//! by choosing a [`StopMode`](crate::compose::StopMode) with [`DockerComposeBuilder::stop_with`](crate::compose::DockerComposeBuilder::stop_with)
+//!
+//! For debugging purpose, you can show `docker compose` output with [`DockerComposeBuilder::inherit_io`](crate::compose::DockerComposeBuilder::inherit_io)
+//!
+//! You can found more examples in the `tests/compose.rs` file
+//!
+
+use std::{
+    collections::HashMap,
+    io::{BufRead, Read},
+    net::SocketAddr,
+    path::{Path, PathBuf},
+    process::{Child, Command, Stdio},
+    sync::Mutex,
+    thread::sleep,
+    time::{Duration, SystemTime},
+};
+
+use bollard_stubs::models::{ContainerStateStatusEnum, HealthStatusEnum};
+use log::{debug, info, trace};
+use once_cell::sync::Lazy;
+use serde::{Deserialize, Serialize};
+
+use crate::core::{
+    logs::{LogStream, WaitError},
+    WaitFor,
+};
+
+mod builder;
+pub use self::builder::*;
+
+/// Define the behavior of the [`DockerCompose`] drop
+/// By default we use [`StopMode::Stop`]
+#[derive(Debug, Clone, Copy)]
+pub enum StopMode {
+    /// Stop containers
+    Stop,
+    /// Stop and remove containers
+    StopAndRemove,
+    /// Do nothing, containers still running after the drop
+    Detach,
+}
+
+impl Default for StopMode {
+    fn default() -> Self {
+        Self::Stop
+    }
+}
+
+/// Implementation of the Docker compose client using the CLI
+
+#[derive(Debug)]
+pub struct DockerCompose {
+    path: PathBuf,
+    env: HashMap<String, String>,
+    child: Option<Child>,
+    ready_conditions: Vec<(String, WaitFor)>,
+    stop_mode: StopMode,
+    inherit_io: bool,
+}
+
+// Trick to avoid code to be run concurrently
+static NO_PARALLEL: Lazy<Mutex<()>> = Lazy::new(Mutex::default);
+
+impl DockerCompose {
+    /// Create the builder
+    pub fn builder(path: impl AsRef<Path>) -> DockerComposeBuilder {
+        DockerComposeBuilder::new(path)
+    }
+
+    fn cmd(&self) -> Command {
+        let mut cmd = Command::new("docker");
+        cmd.arg("compose");
+        cmd.arg("--file");
+        cmd.arg(&self.path);
+        cmd.envs(&self.env);
+
+        cmd
+    }
+
+    fn pull(&self) -> Result<(), std::io::Error> {
+        let mut cmd = self.cmd();
+        cmd.args(["pull"]);
```

Comment:
In the case of not `self.inherit_io`, we need to prevent logs like it's done in the `up()`